statistics/row_sampler.go

```diff
@@ -250,28 +269,44 @@ func (s *baseCollector) ToProto() *tipb.RowSampleCollector {
 	return collector
 }
 
-func (s *baseCollector) FromProto(pbCollector *tipb.RowSampleCollector) {
+func (s *baseCollector) FromProto(pbCollector *tipb.RowSampleCollector, memTracker *memory.Tracker) {
 	s.Count = pbCollector.Count
 	s.NullCount = pbCollector.NullCounts
 	s.FMSketches = make([]*FMSketch, 0, len(pbCollector.FmSketch))
 	for _, pbSketch := range pbCollector.FmSketch {
 		s.FMSketches = append(s.FMSketches, FMSketchFromProto(pbSketch))
 	}
 	s.TotalSizes = pbCollector.TotalSize
-	s.Samples = make(WeightedRowSampleHeap, 0, len(pbCollector.Samples))
+	sampleNum := len(pbCollector.Samples)
+	s.Samples = make(WeightedRowSampleHeap, 0, sampleNum)
+	// consume mandatory memory at the beginning, if exceeds, fast fail
+	if len(pbCollector.Samples) > 0 {
+		rowLen := len(pbCollector.Samples[0].Row)
+		// 24 is the size of datum array, 8 is the size of reference
+		initMemSize := int64(sampleNum) * (int64(rowLen)*types.EmptyDatumSize + EmptyReservoirSampleItemSize + 8)
+		s.MemSize += initMemSize
+		memTracker.Consume(initMemSize)
+	}
+	bufferedMemSize := int64(0)
 	for _, pbSample := range pbCollector.Samples {
-		data := make([]types.Datum, 0, len(pbSample.Row))
+		rowLen := len(pbCollector.Samples[0].Row)
+		data := make([]types.Datum, 0, rowLen)
+		deltaSize := int64(0)
 		for _, col := range pbSample.Row {
 			b := make([]byte, len(col))
 			copy(b, col)
 			data = append(data, types.NewBytesDatum(b))
+			deltaSize += int64(cap(b))
 		}
 		// Directly copy the weight.
-		s.Samples = append(s.Samples, &ReservoirRowSampleItem{
-			Columns: data,
-			Weight:  pbSample.Weight,
-		})
+		sampleItem := &ReservoirRowSampleItem{Columns: data, Weight: pbSample.Weight}
+		s.Samples = append(s.Samples, sampleItem)
+		deltaSize += sampleItem.MemUsage() - EmptyReservoirSampleItemSize
```

Comment:
Do we need to subtract `int64(rowLen)*types.EmptyDatumSize` here? Cuz we pre-consume that at L286

Reply:
You are right, updated.